File: core/src/models/hardware.rs
//! POJO mirrors of the wire-format hardware types defined in
//! `src-tauri/src/models/hardware.rs`.
//!
//! These types are returned by the Core platform sensors. The wire types
//! (with `specta::Type` for tauri-specta TypeScript bindings) live in
//! `src-tauri`; commands and services convert between the two via `From`
//! impls at the boundary.
//!
//! Most types derive `Serialize` / `Deserialize` because the Linux
//! platform layer caches some of them to disk via
//! `crate::platform::linux::cache` (JSON round-trip).

use serde::{Deserialize, Serialize};

use crate::enums::hardware::DiskKind;
use crate::utils::formatter::SizeUnit;

/// `#[serde(rename_all = "camelCase")]` matches the format the Linux
/// platform layer wrote to its on-disk cache before this type moved
/// into Core, so existing cache files stay readable.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemoryInfo {
  pub size: String,
  pub clock: u32,
  pub clock_unit: String,
  pub memory_count: u32,
  pub total_slots: u32,
  pub memory_type: String,
  pub is_detailed: bool,
}

#[derive(Debug, Clone, PartialEq)]
pub struct GraphicInfo {
  pub id: String,
  pub name: String,
  pub vendor_name: String,
  pub clock: u32,
  pub memory_size: String,
  pub memory_size_dedicated: String,
  pub core_count: Option<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct GpuMemoryUsage {
  pub in_use_bytes: Option<String>,
  pub alloc_bytes: Option<String>,
}

/// GPU usage percentage together with the data-source identifier
/// (e.g. "NVAPI", "ADL", "WMI", "DRM (AMD)", "IOKit").
#[derive(Debug, Clone, PartialEq)]
pub struct GpuUsageResult {
  pub usage: i32,
  pub source: String,
}

/// Generic `(name, integer)` pair. Used by the GPU temperature API,
/// where `value` is always in raw degrees Celsius — presentation
/// conversion (Celsius/Fahrenheit) lives at the App-side boundary.
#[derive(Debug, Clone, PartialEq)]
pub struct NameValue {
  pub name: String,
  pub value: i32,
}

#[derive(Debug, Clone, PartialEq)]
pub struct StorageInfo {
  pub name: String,
  pub size: f32,
  pub size_unit: SizeUnit,
  pub free: f32,
  pub free_unit: SizeUnit,
  pub storage_type: DiskKind,
  pub file_system: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct NetworkInfo {
  pub description: Option<String>,
  pub mac_address: Option<String>,
  pub ipv4: Vec<String>,
  pub ipv6: Vec<String>,
  pub link_local_ipv6: Vec<String>,
  pub ip_subnet: Vec<String>,
  pub default_ipv4_gateway: Vec<String>,
  pub default_ipv6_gateway: Vec<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ProcessInfo {
  pub pid: i32,
  pub name: String,
  pub cpu_usage: f32,
  pub memory_usage: f32,
}

#[derive(Debug, Clone, PartialEq)]
pub struct MotherboardInfo {
  pub manufacturer: String,
  pub product: String,
  pub version: String,
  pub serial_number: String,
  pub bios_vendor: String,
  pub bios_version: String,
  pub bios_release_date: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SuperIoChipIdDiagnostics {
  pub platform_supported: bool,
  pub pawnio: Option<PawnIoRuntimeDiagnostics>,
  pub slots: Vec<SuperIoChipIdSlotProbe>,
  pub error: Option<String>,
}

impl SuperIoChipIdDiagnostics {
  /// Result for platforms that do not have a Super I/O `LpcIO` path.
  ///
  /// Reports `platform_supported = false` with an explanatory message
  /// instead of an error, so callers can branch on support uniformly.
  pub fn unsupported_platform() -> Self {
    Self {
      platform_supported: false,
      pawnio: None,
      slots: Vec::new(),
      error: Some(
        "Super I/O LpcIO diagnostics are available on Windows only".to_string(),
      ),
    }
  }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PawnIoRuntimeDiagnostics {
  pub install_location: Option<String>,
  pub dll_path: Option<String>,
  pub module_path: Option<String>,
  pub pawnio_available: bool,
  pub library_loadable: bool,
  pub driver_openable: bool,
  pub module_loadable: bool,
  pub version: Option<u32>,
  pub fallback_reason: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SuperIoChipIdSlotProbe {
  pub slot: u8,
  pub index_port: u16,
  pub data_port: u16,
  pub attempts: Vec<SuperIoChipIdAttempt>,
  pub error: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SuperIoChipIdAttempt {
  pub vendor: SuperIoVendor,
  pub id_high: Option<u8>,
  pub id_low: Option<u8>,
  pub chip_id: Option<u16>,
  pub absent: bool,
  pub error: Option<String>,
  pub exit_error: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SuperIoVendor {
  Nuvoton,
  Ite,
}

#[derive(Debug, Clone, PartialEq)]
pub struct CpuInfo {
  pub name: String,
  pub vendor: String,
  pub core_count: u32,
  pub clock: u32,
  pub clock_unit: String,
  pub cpu_name: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct SysInfo {
  pub cpu: Option<CpuInfo>,
  pub memory: Option<MemoryInfo>,
  pub gpus: Option<Vec<GraphicInfo>>,
  pub storage: Vec<StorageInfo>,
  pub motherboard: Option<MotherboardInfo>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum SmartHealthStatus {
  Passed,
  Failed,
  Unknown,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SmartAttribute {
  pub id: Option<u32>,
  pub name: String,
  pub current: Option<u64>,
  pub worst: Option<u64>,
  pub threshold: Option<u64>,
  pub raw_value: Option<String>,
  pub when_failed: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SmartDiskInfo {
  pub device_name: String,
  pub device_type: Option<String>,
  pub protocol: Option<String>,
  pub model_name: Option<String>,
  pub serial_number: Option<String>,
  pub firmware_version: Option<String>,
  pub capacity_bytes: Option<u64>,
  pub health_status: SmartHealthStatus,
  pub temperature_celsius: Option<i32>,
  pub power_on_hours: Option<u64>,
  pub power_cycle_count: Option<u64>,
  pub attributes: Vec<SmartAttribute>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum StorageHealthStatus {
  Good,
  Warning,
  Critical,
  Unknown,
}

impl StorageHealthStatus {
  pub fn as_str(&self) -> &'static str {
    match self {
      StorageHealthStatus::Good => "good",
      StorageHealthStatus::Warning => "warning",
      StorageHealthStatus::Critical => "critical",
      StorageHealthStatus::Unknown => "unknown",
    }
  }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum StorageWarningLevel {
  None,
  Warning,
  Critical,
  Unknown,
}

impl StorageWarningLevel {
  pub fn as_str(&self) -> &'static str {
    match self {
      StorageWarningLevel::None => "none",
      StorageWarningLevel::Warning => "warning",
      StorageWarningLevel::Critical => "critical",
      StorageWarningLevel::Unknown => "unknown",
    }
  }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StorageHealthRecordDraft {
  pub device_id: String,
  pub date: String,
  pub health_status: StorageHealthStatus,
  pub warning_level: StorageWarningLevel,
  pub temperature_celsius: Option<f32>,
  pub power_on_hours: Option<u64>,
  pub percentage_used: Option<f32>,
  pub available_spare_percent: Option<f32>,
  pub reallocated_sector_count: Option<u64>,
  pub current_pending_sector_count: Option<u64>,
  pub offline_uncorrectable_count: Option<u64>,
  pub media_errors: Option<u64>,
  pub error_log_entries: Option<u64>,
  pub unsafe_shutdown_count: Option<u64>,
  pub warning_reasons: Vec<String>,
  pub collected_at: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StorageHealthRecord {
  pub device_id: String,
  pub display_name: String,
  pub model: Option<String>,
  pub protocol: Option<String>,
  pub capacity_bytes: Option<u64>,
  pub date: String,
  pub health_status: StorageHealthStatus,
  pub warning_level: StorageWarningLevel,
  pub temperature_celsius: Option<f32>,
  pub power_on_hours: Option<u64>,
  pub percentage_used: Option<f32>,
  pub available_spare_percent: Option<f32>,
  pub reallocated_sector_count: Option<u64>,
  pub current_pending_sector_count: Option<u64>,
  pub offline_uncorrectable_count: Option<u64>,
  pub media_errors: Option<u64>,
  pub error_log_entries: Option<u64>,
  pub unsafe_shutdown_count: Option<u64>,
  pub warning_reasons: Vec<String>,
  pub collected_at: String,
}

/// One device's Live Storage Health signals (ADR 0006).
///
/// Collected on demand for immediate display and never persisted,
/// unlike the daily [`StorageHealthRecord`].
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LiveStorageHealth {
  /// Same HMAC-based identity as the daily Storage Health Record, so
  /// displays can join live signals to persisted device rows.
  pub device_id: String,
  pub display_name: String,
  pub temperature_celsius: Option<f32>,
  pub available_spare_percent: Option<f32>,
  pub percentage_used: Option<f32>,
  pub media_errors: Option<u64>,
  pub unsafe_shutdown_count: Option<u64>,
  pub power_on_hours: Option<u64>,
  pub power_cycle_count: Option<u64>,
  pub collected_at: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StorageDeviceRecord {
  pub id: String,
  pub display_name: String,
  pub model: Option<String>,
  pub serial_hash: Option<String>,
  pub protocol: Option<String>,
  pub capacity_bytes: Option<u64>,
  pub first_seen_at: String,
  pub last_seen_at: String,
}

File: src/e2e/fixtures/hardware.ts
import type {
  HardwareMonitorUpdate,
  ProcessInfo_Serialize,
  StorageHealthRecord,
  StorageInfo,
  SysInfo,
} from "@/rspc/bindings";

/**
 * GPU ids must match between `sysInfoFixture.gpus[].id` and the
 * `hardware-monitor-update` payloads (`gpus[].gpuId`) — the dashboard joins
 * them to resolve the selected GPU. Two GPUs are provided so the GPU selector
 * tablist renders and can be driven via accessible selectors.
 */
export const GPU_FIXTURES = [
  { id: "e2e-gpu-0", name: "HV Fixture GPU 8GB" },
  { id: "e2e-gpu-1", name: "HV Fixture iGPU" },
] as const;

export const sysInfoFixture: SysInfo = {
  cpu: {
    name: "HV Fixture CPU 8-Core",
    vendor: "FixtureVendor",
    coreCount: 8,
    clock: 3600,
    clockUnit: "MHz",
    cpuName: "HV Fixture CPU 8-Core",
  },
  memory: {
    size: "32 GB",
    clock: 4800,
    clockUnit: "MHz",
    memoryCount: 2,
    totalSlots: 4,
    memoryType: "DDR5",
    isDetailed: false,
  },
  gpus: [
    {
      id: GPU_FIXTURES[0].id,
      name: GPU_FIXTURES[0].name,
      vendorName: "FixtureVendor",
      clock: 2100,
      memorySize: "8 GB",
      memorySizeDedicated: "8 GB",
      coreCount: null,
    },
    {
      id: GPU_FIXTURES[1].id,
      name: GPU_FIXTURES[1].name,
      vendorName: "FixtureVendor",
      clock: 1500,
      memorySize: "2 GB",
      memorySizeDedicated: "2 GB",
      coreCount: null,
    },
  ],
  storage: [
    {
      name: "Fixture SSD",
      size: 953,
      sizeUnit: "GB",
      free: 512,
      freeUnit: "GB",
      storageType: "ssd",
      fileSystem: "NTFS",
    },
    {
      name: "Fixture HDD",
      size: 3,
      sizeUnit: "GB",
      free: 1,
      freeUnit: "GB",
      storageType: "hdd",
      fileSystem: "NTFS",
    },
  ],
  motherboard: {
    manufacturer: "FixtureVendor",
    product: "HV Fixture Board",
    version: "1.0",
    serialNumber: "E2E-0000",
    biosVendor: "FixtureVendor",
    biosVersion: "1.0.0",
    biosReleaseDate: "2024-01-01",
  },
};

export const processListFixture: ProcessInfo_Serialize[] = [
  { pid: 100, name: "hv-fixture-app", cpuUsage: "12.5", memoryUsage: "256" },
  { pid: 200, name: "fixture-browser", cpuUsage: "8.1", memoryUsage: "1228.8" },
  { pid: 300, name: "fixture-editor", cpuUsage: "4.4", memoryUsage: "512" },
  { pid: 400, name: "fixture-daemon", cpuUsage: "1.2", memoryUsage: "64" },
  { pid: 500, name: "fixture-shell", cpuUsage: "0.3", memoryUsage: "32" },
];

export const buildStorageInfoFixture = (count: number): StorageInfo[] =>
  Array.from({ length: Math.max(0, count) }, (_, index) => {
    if (index === 0) {
      return {
        name: "Fixture SSD",
        size: 953,
        sizeUnit: "GB",
        free: 512,
        freeUnit: "GB",
        storageType: "ssd",
        fileSystem: "NTFS",
      };
    }

    if (index === 1) {
      return {
        name: "Fixture HDD",
        size: 3,
        sizeUnit: "GB",
        free: 1,
        freeUnit: "GB",
        storageType: "hdd",
        fileSystem: "NTFS",
      };
    }

    const size = index % 4 === 0 ? 4 : 1;
    const storageType = index % 5 === 0 ? "hdd" : "ssd";

    return {
      name: `Fixture ${storageType.toUpperCase()} ${index + 1}`,
      size,
      sizeUnit: "GB",
      free: Math.max(1, size - 1),
      freeUnit: "GB",
      storageType,
      fileSystem: index % 3 === 0 ? "exFAT" : "NTFS",
    };
  });

export const buildStorageHealthFixture = (
  count: number,
  options: {
    collectedAt?: string;
    date?: string;
  } = {},
): StorageHealthRecord[] =>
  Array.from({ length: Math.max(0, count) }, (_, index) => {
    const date = options.date ?? "2026-01-01";
    const isCritical = index > 0 && index % 7 === 0;
    const isWarning = !isCritical && index > 0 && index % 3 === 0;
    const healthStatus = isCritical
      ? "critical"
      : isWarning
        ? "warning"
        : "good";
    const displayName =
      index === 0
        ? "Fixture SSD"
        : index === 1
          ? "Fixture HDD"
          : `Fixture ${index % 5 === 0 ? "HDD" : "SSD"} ${index + 1}`;
    const temperatureCelsius = 36 + ((index * 3) % 29);
    const percentageUsed = Math.min(99, 3 + index * 4);
    const warningReasons = isCritical
      ? [
          `Fixture disk ${index + 1} has critical media errors`,
          `Temperature reached ${temperatureCelsius} C`,
        ]
      : isWarning
        ? [`Fixture disk ${index + 1} is above the warning threshold`]
        : [];

    return {
      deviceId: `e2e-disk-${index}`,
      displayName,
      model:
        index === 0
          ? "FIXTURE-SSD-1TB"
          : `FIXTURE-${index % 5 === 0 ? "HDD" : "SSD"}-${String(index + 1).padStart(2, "0")}`,
      protocol: index % 5 === 0 ? "SATA" : "NVMe",
      capacityBytes: (index % 4 === 0 ? 4 : 1) * 1_024_000_000_000,
      date,
      healthStatus,
      warningLevel: healthStatus === "good" ? "none" : healthStatus,
      temperatureCelsius,
      powerOnHours: 1234 + index * 911,
      percentageUsed,
      availableSparePercent: Math.max(1, 100 - index * 3),
      reallocatedSectorCount: isCritical ? 12 + index : null,
      currentPendingSectorCount: isWarning || isCritical ? index : null,
      offlineUncorrectableCount: isCritical ? 2 : null,
      mediaErrors: isCritical ? 5 + index : 0,
      errorLogEntries: isWarning || isCritical ? index * 2 : 0,
      unsafeShutdownCount: 2 + index,
      warningReasons,
      collectedAt: options.collectedAt ?? `${date}T00:00:00Z`,
    };
  });

export const storageHealthFixture: StorageHealthRecord[] =
  buildStorageHealthFixture(1);

/**
 * Build a deterministic series of `hardware-monitor-update` payloads.
 * Values follow fixed sine waves so charts always render the same shape.
 */
export const buildHardwareUpdateSeries = (
  length: number,
): HardwareMonitorUpdate[] =>
  Array.from({ length }, (_, i) => ({
    cpuUsage: round1(45 + 25 * Math.sin(i / 4)),
    memoryUsage: round1(62 + 6 * Math.sin(i / 6 + 1)),
    gpus: [
      {
        gpuId: GPU_FIXTURES[0].id,
        gpuName: GPU_FIXTURES[0].name,
        gpuUsage: round1(55 + 35 * Math.sin(i / 3 + 2)),
        gpuTemperature: round1(58 + 6 * Math.sin(i / 5)),
        gpuSource: "fixture",
        gpuDedicatedMemoryUsageKb: 4_194_304,
        gpuCoolerLevel: 42,
      },
      {
        gpuId: GPU_FIXTURES[1].id,
        gpuName: GPU_FIXTURES[1].name,
        gpuUsage: round1(20 + 10 * Math.sin(i / 4 + 1)),
        gpuTemperature: round1(45 + 4 * Math.sin(i / 5 + 1)),
        gpuSource: "fixture",
        gpuDedicatedMemoryUsageKb: 1_048_576,
        gpuCoolerLevel: 30,
      },
    ],
    processorsUsage: Array.from({ length: 8 }, (_, core) =>
      round1(40 + 30 * Math.sin(i / 4 + core)),
    ),
    cpuTemperature: Math.round(50 + 8 * Math.sin(i / 5)),
    sensorTemperatures: [
      { name: "CPUZ", value: Math.round(50 + 8 * Math.sin(i / 5)) },
      { name: "TZ01", value: Math.round(42 + 5 * Math.sin(i / 6)) },
    ],
    motherboardTemperatures: [
      {
        name: "SYSTIN",
        value: Math.round(38 + 4 * Math.sin(i / 6)),
        source: "NCT6799D / Super I/O",
      },
      {
        name: "CPUTIN",
        value: Math.round(48 + 8 * Math.sin(i / 5)),
        source: "NCT6799D / Super I/O",
      },
      {
        name: "AUXTIN0",
        value: Math.round(35 + 3 * Math.sin(i / 7)),
        source: "NCT6799D / Super I/O",
      },
    ],
    motherboardFanSpeeds: [
      {
        name: "Fan 1",
        rpm: Math.round(1200 + 120 * Math.sin(i / 4)),
        status: "active",
        source: "NCT6799D / Super I/O",
      },
      {
        name: "Fan 2",
        rpm: 0,
        status: "inactive",
        source: "NCT6799D / Super I/O",
      },
    ],
  }));

const round1 = (value: number) => Math.round(value * 10) / 10;
